tests/api-tests/access-control/authed.test.js

```diff
@@ -164,17 +162,21 @@ multiAdapterRunners().map(({ before, after, adapterName }) =>
 
     describe('read', () => {
       test('authed user', async () => {
-        const query = `query { authenticatedUser { id yesRead noRead } }`;
-        const context = keystone.createContext({ authentication: { item: user, listKey: 'User' } });
-        const { data, errors } = await keystone.executeGraphQL({ context, query });
-        expect(data.authenticatedUser).not.toBe(null);
-        expect(data.authenticatedUser.id).toEqual(user.id);
-        expect(data.authenticatedUser.yesRead).toEqual(user.yesRead);
-        expect(data.authenticatedUser.noRead).toEqual(null);
+        const query = `query { authenticatedItem { id yesRead noRead } }`;
+        const _context = context.exitSudo().withSession({
+          itemId: user.id,
+          listKey: 'User',
+          data: user,
+        });
+        const { data, errors } = await _context.executeGraphQL({ query });
+        expect(data.authenticatedItem).not.toBe(null);
+        expect(data.authenticatedItem.id).toEqual(user.id);
+        expect(data.authenticatedItem.yesRead).toEqual(user.yesRead);
+        expect(data.authenticatedItem.noRead).toEqual(null);
         expect(errors).toHaveLength(1);
         expect(errors[0].name).toEqual('GraphQLError');
         expect(errors[0].message).toEqual('You do not have access to this resource');
-        expect(errors[0].path).toEqual(['authenticatedUser', 'noRead']);
+        expect(errors[0].path).toEqual(['authenticatedItem', 'noRead']);
```

Comment:
The fact that this test passes worries me and I would guess that there's a bug in this test. It should fail with a GraphQL validation error because authenticatedItem returns `AuthenticatedItem` which is a union so to access fields on the user, you need to write `... on User { id yesRead noRead }`. Could you see what happens if you change the query to use `... on User {id yesRead noRead}` and figure out why it's not failing with a validation error currently?

Reply:
It seems that the Apollo is responsible for making the validation error that you're expecting. Running the same query in the graphql playground gives:

```{
  "error": {
    "errors": [
      {
        "message": "Cannot query field \"id\" on type \"AuthenticatedItem\". Did you mean to use an inline fragment on \"User\"?",
        "locations": [
          {
            "line": 3,
            "column": 5
          }
        ],
        "extensions": {
          "code": "GRAPHQL_VALIDATION_FAILED",
          "exception": {
            "stacktrace": [
              "GraphQLError: Cannot query field \"id\" on type \"AuthenticatedItem\". Did you mean to use an inline fragment on \"User\"?",
              "    at Object.Field (/Users/timleslie/src/keystone-5/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:48:31)",
              "    at Object.enter (/Users/timleslie/src/keystone-5/node_modules/graphql/language/visitor.js:323:29)",
              "    at Object.enter (/Users/timleslie/src/keystone-5/node_modules/graphql/utilities/TypeInfo.js:370:25)",
              "    at visit (/Users/timleslie/src/keystone-5/node_modules/graphql/language/visitor.js:243:26)",
              "    at Object.validate (/Users/timleslie/src/keystone-5/node_modules/graphql/validation/validate.js:69:24)",
              "    at validate (/Users/timleslie/src/keystone-5/node_modules/apollo-server-core/src/requestPipeline.ts:514:14)",
              "    at Object.<anonymous> (/Users/timleslie/src/keystone-5/node_modules/apollo-server-core/src/requestPipeline.ts:300:32)",
              "    at Generator.next (<anonymous>)",
              "    at fulfilled (/Users/timleslie/src/keystone-5/node_modules/apollo-server-core/dist/requestPipeline.js:5:58)",
              "    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
            ]
          }
        },
        "uid": "ckm5meu3x0002rimuaa0a4pq7",
        "name": "ValidationError"
      }
    ]
  }
}```

The test framework doesn't go through apollo however, it directly calls `execute` against the GraphQL schema. I suspect this explains the discrepancy between the two cases. I've updated the tests to use the more correct inline fragment.